text/0000-generic-int.md

```diff
@@ -0,0 +1,552 @@
+would get optimised to the equivalent:
+
+```rust
+fn count_zeros(self) -> u32 {
+    (self as u8).count_zeros() - 1
+}
+```
+
+For the most part, the code which uses intrinsics will be specialised, and the
+code which doesn't will be replaced with a generic version. For example, `ctpop`
+is required for `count_zeros`, and thus it's specialized. However, `Default`
+does not require any specialisation, and can just be replaced with a single
+generic impl.
+
+## Documentation
+
+For now, additional `primitive.uint` and `primitive.int` pages will be added to
+rustdoc, and the existing `uN` and `iN` pages will be left as-is. Eventually, if
+we're comfortable with it, we can remove the `uN` and `iN` pages entirely and
+use `primitive.int` and `primitive.uint` as the single source of documentation
+for these types, plus the pages for `usize` and `isize`.
+
+There certainly is a precedent for this: as of right now, all of these pages
+share the same documentation, and the examples are modified to work for all of
+these types. Removing these separate pages would help remove documentation
+redundancy, although `usize` and `isize` would still have to be kept separate.
+
+# Potential future extensions
+
+## `N > 128`
+
+The most obvious extension is to allow arbitrarily sized integer types. One
+clear use for this is cryptographic code which often works with
+fixed-but-large-length integers, which would no longer have to decide whether
+to offer its own `u256` or `u4096` types. Additionally, generalisation over
+integers could allow code that works with e.g. 1024-, 2048-, and 4096-bit keys
+to have their arithmetic monomorphised and specialised for free without writing
+extra code.
+
+Another benefit of this would be allowing arbitrary-length integer literals,
+which would be useful for bignum libraries. Instead of parsing a string or byte
+array to create a bignum, the bignum library could simply have:
+
+```
+fn new<const N: usize>(num: uint<N>) -> BigNum {
+    // ...
+}
+```
+
+Internally, the compiler could represent the power-of-two `uN` and `iN` as a
+`doubled_uint<N / 2>` and `doubled_int<N / 2>`, which could even be
+implemented in the standard library. At that point, only a lang item would be
+necessary for allowing literal coercion.
+
+Of course, at this point, how this extension would actually work is complete
+speculation. It seems inevitable with this RFC, although due to the large number
+of unsolved problems, should require one or several more RFCs to stabilise.
+
+## Bit sizes and `repr(bitpacked)`
+
+In the future, types could be sized in terms of bits instead of bytes, with
+`(bit_size_of::<T>() + 7) / 8 == size_of::<T>()`. All types would have a bit
+size, allowing for a future `repr(bitpacked)` extension which packs all values
+in a struct or enum variant into the smallest number of bytes possible, given
+their bit sizes. Doing so would prevent referencing the fields of the struct,
```

Comment:
Since you are currently limiting the size to 128 bits (but maybe more in the future) wouldn't it be possible to define a reference to a bitpacked generic integer as a `(ref, (u16, u16))` where the second pair is a (start, length) pair within?